### ISSUE TITLE
CIFAR data preparation

### ISSUE THREAD

**Issue by danpovey** [kaldi-asr/kaldi#1549]
@hhadian, I was hoping that you could help with this since I believe you have a CIFAR setup.

This is a request for you to do the data preparation for CIFAR-10 and CIFAR-100, e.g. as egs/cifar/v1 (we can have one directory for both).  I'm not calling it s5 since this isn't a speech setup, for non-speech setups we're starting with v1.

Let's have a directories data/cifar10_train and data/cifar10_test (and the same for cifar100).
Each directory will have files:
  labels.txt
images.scp

where labels.txt will just be something like:
00001 3
00002 0
00003 9
and so on, with the labels, and you can have a file classes.txt (like words.txt) explaining what the labels mean, like
boat 0
car 1
and so on.
The images.scp will point to compressed feature matrices in archives somewhere in a subdirectory data/.  I'm working on modifications of the compression code so we can make it lossless in this case, and I'll give you details later on.

There will be a lot more on this thread, regarding creating the egs and so on, but I think this is enough to get you started.

**Comment by hhadian** [comment]
OK, sure I will do it.

**Comment by hhadian** [comment]
I'm a bit confused. Do you want me to create a script e.g. data_prep.sh (similar to ASR recipes) that creates these files (i.e. labels.txt, images.scp, classes.txt) for CIFAR or do you want me to actually create these files in the data directory?

**Comment by danpovey** [comment]
Create the script that downloads the CIFAR data and creates those files.
I'll upload, in a minute, a validation script that I've been working on.

**Comment by hhadian** [comment]
OK, that's what I thought.

**Comment by danpovey** [comment]
Also, can you share what the configuration you are using in Keras for CIFAR
image augmentation is?  They seem to have a class ImageDataGenerator that
takes a bunch of options.  I am wondering about the best way to do image
augmentation in Kaldi.  Keras seems like a very big dependency to take on
if we just need the image augmentation, since it would require installing
Tensorflow or Theano.  I'm wondering if we can replicate it in our own
python scripts.

Incidentally, I don't necessarily expect that we will get world-beating
results in image classification; part of the reason I want to make this
setup is to make it easier to publish nnet3-based results in places like
NIPS where multiple tasks are required; and another reason is that many of
the techniques and ideas will be reusable for text and handwriting
recognition.

**Comment by hhadian** [comment]
Yes, ImageDataGenerator in Keras offers quite a lot of options. In the recipe I am using only 3 options are enabled: vertical/horizontal shift (by 10%) and horizontal flipping. I have not tried the rest of them, but since the current relative improvement due to data augmentation (~44%) is similar to a reported CIFAR paper, I guess the other options won't help much probably [at least for CIFAR-10].

Anyway, ImageDataGenerator is implemented in a preprocessing module in Keras and depends on numpy and scipy. If we can depend on scipy/numpy we will probably be able to use it (I don't know about licensing though). The module is [here](https://github.com/hhadian/keras/blob/master/keras/preprocessing/image.py).

BTW, all of my changes to Keras for backstitch are [here](https://github.com/hhadian/keras/commit/55ccaa6c0a63dad70e6165928af36d9f17b3e679) if you ever want to take a look.
